File: rb/test/find_entity_test.rb
# Find entity test

require "minitest/autorun"
require "json"
require_relative "../Anipub_sdk"
require_relative "runner"

class FindEntityTest < Minitest::Test
  def test_create_instance
    testsdk = AnipubSDK.test(nil, nil)
    ent = testsdk.Find(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = find_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "find." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set ANIPUB_TEST_FIND_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    find_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.find")))
    find_ref01_data = nil
    if find_ref01_data_raw.length > 0
      find_ref01_data = Helpers.to_map(find_ref01_data_raw[0][1])
    end

    # LOAD
    find_ref01_ent = client.Find(nil)
    find_ref01_match_dt0 = {
      "id" => find_ref01_data["id"],
    }
    find_ref01_data_dt0_loaded = find_ref01_ent.load(find_ref01_match_dt0, nil)
    find_ref01_data_dt0_load_result = Helpers.to_map(find_ref01_data_dt0_loaded.respond_to?(:data_get) ? find_ref01_data_dt0_loaded.data_get : find_ref01_data_dt0_loaded)
    assert !find_ref01_data_dt0_load_result.nil?
    assert_equal find_ref01_data_dt0_load_result["id"], find_ref01_data["id"]

  end
end

def find_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "find", "FindTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = AnipubSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["find01", "find02", "find03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["ANIPUB_TEST_FIND_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "ANIPUB_TEST_FIND_ENTID" => idmap,
    "ANIPUB_TEST_LIVE" => "FALSE",
    "ANIPUB_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["ANIPUB_TEST_FIND_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["ANIPUB_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = AnipubSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["ANIPUB_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["ANIPUB_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: ts/test/entity/find/FindEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { AnipubSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('FindEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when ANIPUB_TEST_LIVE=TRUE.
  afterEach(liveDelay('ANIPUB_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = AnipubSDK.test()
    const ent = testsdk.Find()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.ANIPUB_TEST_LIVE
    for (const op of ['load']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'find.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"ep","req":false,"short":"Episode count if found","type":"`$INTEGER`","index$":0},{"active":true,"name":"exist","req":true,"short":"Whether anime exists","type":"`$BOOLEAN`","index$":1},{"active":true,"name":"id","req":false,"short":"Anime ID if found","type":"`$INTEGER`","index$":2}],"id":{"field":"id","name":"id"},"name":"find","op":{"load":{"input":"data","name":"load","points":[{"active":true,"args":{"params":[{"active":true,"example":"One Piece","kind":"param","name":"id","orig":"name","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /api/find/{name}","json":"{\"operationId\":\"findAnimeByName\",\"parameters\":[{\"description\":\"URL-encoded anime name — spaces as %20\",\"example\":\"One Piece\",\"in\":\"path\",\"name\":\"name\",\"required\":true,\"schema\":{\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"examples\":{\"found\":{\"value\":{\"ep\":1155,\"exist\":true,\"id\":10}},\"notFound\":{\"value\":{\"exist\":false}}},\"schema\":{\"properties\":{\"ep\":{\"description\":\"Episode count if found\",\"type\":\"integer\"},\"exist\":{\"description\":\"Whether anime exists\",\"type\":\"boolean\"},\"id\":{\"description\":\"Anime ID if found\",\"type\":\"integer\"}},\"required\":[\"exist\"],\"type\":\"object\"}}},\"description\":\"Anime found or not found response\"},\"500\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"error\":{\"description\":\"Error message\",\"type\":\"string\"},\"status\":{\"description\":\"HTTP status code\",\"type\":\"integer\"}},\"type\":\"object\"}}},\"description\":\"Internal server error\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/api/find/{name}","rename":{"param":{"name":"id"}},"segments":[{"lit":"api"},{"lit":"find"},{"var":"id"}],"select":{"exist":["id"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":0}],"key$":"load"}},"relations":{"ancestors":[]},"key$":"find","name__orig":"find","Name":"Find","name_":"find","name-":"find","NAME":"FIND","index$":1}, {"active":true,"entity":"find","key$":"BasicFindFlow","kind":"basic","name":"BasicFindFlow","param":{},"step":[{"active":true,"data":{},"input":{"ref":"find_ref01","srcdatavar":"find_ref01_data","suffix":"_dt0"},"match":{"id":"find01"},"op":"load","spec":[],"valid":[{"apply":"TextFieldMark","def":{"mark":"Mark01-find_ref01"}}],"index$":0}]}, 'Find')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let find_ref01_data = Object.values(setup.data.existing.find)[0] as any

    // LOAD
    const find_ref01_ent = client.Find()
    const find_ref01_match_dt0: any = {}
    find_ref01_match_dt0.id = find_ref01_data.id
    const find_ref01_data_dt0 = (await find_ref01_ent.load(find_ref01_match_dt0)).data()
    assert(find_ref01_data_dt0.id === find_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/find/FindTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = AnipubSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['find01','find02','find03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'ANIPUB_TEST_FIND_ENTID': idmap,
    'ANIPUB_TEST_LIVE': 'FALSE',
    'ANIPUB_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['ANIPUB_TEST_FIND_ENTID']

  const live = 'TRUE' === env.ANIPUB_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['ANIPUB_TEST_FIND_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new AnipubSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.ANIPUB_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  
